### PR TITLE
add prefetching to binned implementations.  Set prefetch size to 50 b…

### PR DESCRIPTION
**Pull request by jbrowne6**
…ased on experiments.

No tests done, but I basically copied code from the base forest types to the binned.

### CODE REVIEW

**Comment by falkben** [comment]
When trying to build on Windows using visual studio build tools:

```cmd
cl /EHsc /openmp src/fp.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.16.27026.1 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

fp.cpp
c:\users\ben\documents\repos\neurodata\rerf\packedforest\src\foresttypes\binnedtree\binStruct.h(342): error C3005: 'update': unexpected token encountered on OpenMP 'atomic' directive
c:\users\ben\documents\repos\neurodata\rerf\packedforest\src\foresttypes\binnedtree\binStruct.h(475): note: see reference to class template instantiation 'fp::binStruct<T,Q>' being compiled
c:\users\ben\documents\repos\neurodata\rerf\packedforest\src\foresttypes\binnedtree\binStruct.h(380): error C3005: 'update': unexpected token encountered on OpenMP 'atomic' directive
c:\users\ben\documents\repos\neurodata\rerf\packedforest\src\foresttypes\binnedtree\binStruct.h(414): error C3005: 'update': unexpected token encountered on OpenMP 'atomic' directive
c:\users\ben\documents\repos\neurodata\rerf\packedforest\src\foresttypes\binnedtree\binStruct.h(451): error C3005: 'update': unexpected token encountered on OpenMP 'atomic' directive
c:\users\ben\documents\repos\neurodata\rerf\packedforest\src\baseFunctions/fpForest.h(11): warning C4273: 'omp_set_dynamic': inconsistent dll linkage
C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\VC\Tools\MSVC\14.16.27023\include\omp.h(66): note: see previous definition of 'omp_set_dynamic'
c:\users\ben\documents\repos\neurodata\rerf\packedforest\src\baseFunctions/fpForest.h(14): warning C4273: 'omp_set_num_threads': inconsistent dll linkage
C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\VC\Tools\MSVC\14.16.27023\include\omp.h(41): note: see previous definition of 'omp_set_num_threads
'```

**Comment by falkben** [comment]
Can we add the evidence for this (just an image)?

**Comment by jbrowne6** [comment]
[prefetch.pdf](https://github.com/neurodata/RerF/files/2957031/prefetch.pdf)

**Comment by jbrowne6** [comment]
I fixed the tests and have tested that prefetching is actually working using g++ on Ubuntu.

File: packedForest/src/fpSingleton/dataset/fpDataBase.h [hunk 1]
```diff
@@ -1,224 +1,225 @@
...
+				inline int returnNumClasses(){
+					return Y.numClasses();
+				}
 
-		void printDataStats(){
-			std::cout << "there are " << this->returnNumFeatures() << " features.\n";
-			std::cout << "there are " << this->returnNumObservations() << " observations.\n";
-		}
+				void printDataStats(){
+					std::cout << "there are " << this.returnNumFeatures() << " features.\n";
```
Comment (falkben) [inline packedForest/src/fpSingleton/dataset/fpDataBase.h]: `this.` no good.

**Pull request merged**